Warn about sending on Sendable type in function parameters (#74616)

Androp0v opened the pull request:
When a function parameter is Sendable, adding 'sending' annotation does nothing, as the value is safe to be sent across isolation boundaries. SE-0430 seems to imply that 'sending' is intended to be used for non-Sendable types only.

Includes fix-it to remove 'sending' from parameter.

Addresses #74616 .

Androp0v committed:
Warn about sending on Sendable type in function parameters (#74616)
When a function parameter is Sendable, adding 'sending' annotation does nothing, as the value is safe to be sent across isolation boundaries. SE-0430 seems to imply that 'sending' is intended to be used for non-Sendable types only.

hborla commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -4975,8 +4975,29 @@ TypeResolver::resolveSendingTypeRepr(SendingTypeRepr *repr,
+  
+  // If resolved type is Sendable, warn about unnecessary 'sending' annotation.
+  if (options.is(TypeResolverContext::FunctionInput)) {
+    if(!resolvedType->hasTypeParameter()) {
```
I think we still want to check this if the type has a type parameter. If we're in the interface stage of type resolution, we can check the generic signature to see if it has a `Sendable` requirement.

Androp0v replied:
Updated the PR using generic signature to check for `Sendable` requirement

hborla commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -4975,8 +4975,29 @@ TypeResolver::resolveSendingTypeRepr(SendingTypeRepr *repr,
+        .fixItRemove(repr->getSpecifierLoc());
+      }
+    } else {
+      if (resolvedType->getASTContext().getProtocol(KnownProtocolKind::Sendable)) {
```
This is only getting the known protocol declaration from the ASTContext. Are you trying to check whether a type with a type parameter is conforming to `Sendable` here?

Androp0v commented:
I'll address your comments and update the PR, thanks @hborla

Androp0v committed:
Use generic signature instead of ASTContext to warn about sending on Sendable type
Follow PR review comments and extend the warnings about sending on sendable types to cover types with a type parameter, and use generic signature instead of ASTContext to check conformance to Sendable. Additionally, use different warning messages based on whether the sending on Sendable is used on a function parameter or function result.

Androp0v commented:
FYI: I didn't add tests for:
```
func testSendingOnSendableResult() -> sending some Sendable { ... }
```
Due to https://github.com/swiftlang/swift/issues/74846
